source/shared/cpp/ObjectModel/ColumnSet.cpp

```diff
@@ -47,11 +47,65 @@ std::shared_ptr<BaseCardElement> ColumnSetParser::Deserialize(ParseContext& cont
 }
 
 void ColumnSet::DeserializeChildren(ParseContext& context, const Json::Value& value)
-{
-    // Parse Columns
-    auto cardElements =
-        ParseUtil::GetElementCollectionOfSingleType<Column>(context, value, AdaptiveCardSchemaKey::Columns, Column::Deserialize, false);
-    m_columns = std::move(cardElements);
+{ 
+    auto elementArray = ParseUtil::GetArray(value, AdaptiveCardSchemaKey::Columns, false);
+
+    std::vector<std::shared_ptr<Column>> elements;
+    if (elementArray.empty())
+    {
+        m_columns = elements;
+        return;
+    }
+
+    const size_t elemSize = elementArray.size();
+    const size_t endElemIndex = elemSize - 1;
+    elements.reserve(elemSize);
+
+    size_t currentIndex = 0;
+
+    // only Lc (LeftChild) & Rc (RightChild) get Padding
+    // lc inherites bleedToLeading from parent who is Lc and bleedToLeading
+    // or parent is BleedToBothEdges
+    // same applies to Rc
+    ContainerBleedState previousBleedState = context.GetBleedState();
+
+    if (elemSize == 1)
+    {
+        // items in signle column has no restriction
+        context.SetBleedState(ContainerBleedState::BleedToBothEdges);
+    }
+
+    // Deserialize every element in the array
+    for (const Json::Value& curJsonValue : elementArray)
+    {
+        // processing the cases where parent's bleed state is not restricted
+        if (elemSize != 1)
+        {
+            if (currentIndex == 0)
+            {
+                context.SetBleedState(ContainerBleedState::BleedToLeading);
+            }
+            else if (currentIndex == endElemIndex)
+            {
+                context.SetBleedState(ContainerBleedState::BleedToTrailing);
+            } else
+            {
+                context.SetBleedState(ContainerBleedState::BleedRestricted);
+            }
+        }
+
+        // Parse the element
+        auto el = Column::Deserialize(context, curJsonValue);
+        if (el != nullptr)
+        {
+            elements.push_back(el);
+            // restores the parent's bleed state
+            context.SetBleedState(previousBleedState);
+        }
+        ++currentIndex;
+    }
```

Comment:
why not just use a regular `for` loop here since you're maintaining the index?

Reply:
sounds great. thanks

---
In reply to: [265289875](https://github.com/Microsoft/AdaptiveCards/pull/2544#discussion_r265289875) [](ancestors = 265289875)